rfd/0048-war-dialler-node-tracker.md

```diff
@@ -0,0 +1,199 @@
+	3- Proxy-1 opens a tunnel (RFD 49?) to proxy-2 and asks proxy-2 to dial the requested node.
+
+	Case 2 - connection 2:
+	nothing changes when a client connects to a proxy that has got a direct connection to the requested node.
+```
+
+### Scope
+
+The first iteration of the node tracker service will be a barebones implementation of an in memory cache represented by a simple golang map.
+
+### How it works
+
+The node tracker service runs a grpc server and exposes three endpoints to teleport proxies:
+- AddNode: which takes a `node ID`, `proxy ID` and `proxy address` as parameters among others. A Proxy will call this endpoint when it receives a heartbeat request from a node agent connected via a reverse tunnel, signaling to the node tracker that the node can be reached through this proxy, thus creating a node-to-proxy relationship.
+- RemoveNode: which takes a `node ID` as a parameter. A proxy will call this endpoint when it fails to receive heartbeat requests from a node agent after a certain threshold.
+- GetProxies: which takes `node ID` as a parameter. A proxy will call this endpoint when looking up which proxy is connected to a node it doesn't have a direct connection to. The node tracker will reply with the addresses of proxies that can reach the node.
+
+
+The service also runs a background job responsible for cleaning stale relationships that haven't been updated before a certain threshold.
+
+### Limits
+
+Currently we only support running one instance of a node tracker and because of that we're missing key features usually found in distributed systems, such as:
+- Scalability
+- High availabitly
+- Consistency
+- Redundancy
+- Consensus
+- Sharding
+
+Persistence is also an issue due to the nature of our in memory design. A reboot or a failure will render the majority of the nodes in the cluster unreachable while the node tracker rebuilds its database.
+
+These will hopefully be implemented in future iterations to this project as we iron out the issues and get a better sense of the requirements.
+
+There are some low hanging fruits that can improve the performance and reliability of the current design:
+- Proxy triggered updates: Proxies can ping their connected nodes in order to trigger an update to the node tracker instead of waiting for heartbeats. This can improve the speed of rebuilding the node tracker database after a reboot.
+- Add a cleanup endpoint to the node tracker to remove all node-to-proxy relationships from the node tracker for cases when a proxy is recycled or becomes unavailable.
+- Caching recent lookups at the proxy level. (Note to David/Kevin: this should be part of the initial implementation)
+
+## Code
+
+### Config
+
+A typical config for the node tracker will look like
+```
+node_tracker_service:
+  enabled: yes
+  listen_addr: ...
+
+  proxy_keep_alive_interval: ... // Ex: 5m. Represents the cleaning threshold for stale node-to-proxy relationships.
+
+  https_keypairs:
```

Comment:
We should talk about if we want to support external certs from the start. I think external certs make sense for the idea of supporting multiple tenants but for a single cluster it makes sense to register with the auth server the same way a proxy does.

Reply:
I think it's an "either or" situation. We'll use external certs if provided or have them generated otherwise.